src/main/scala/cognite/spark/v1/DefaultSource.scala

```diff
@@ -185,49 +190,54 @@ class DefaultSource
       data: DataFrame): BaseRelation = {
     val config = parseRelationConfig(parameters, sqlContext)
     val resourceType = parameters.getOrElse("type", sys.error("Resource type must be specified"))
-    val relation = resourceType match {
-      case "events" =>
-        new EventsRelation(config)(sqlContext)
-      case "timeseries" =>
-        val useLegacyName = toBoolean(parameters, "useLegacyName")
-        new TimeSeriesRelation(config, useLegacyName)(sqlContext)
-      case "assets" =>
-        new AssetsRelation(config)(sqlContext)
-      case "datapoints" =>
-        new NumericDataPointsRelationV1(config)(sqlContext)
-      case "stringdatapoints" =>
-        new StringDataPointsRelationV1(config)(sqlContext)
-      case _ => sys.error(s"Resource type $resourceType does not support save()")
-    }
-
-    data.foreachPartition((rows: Iterator[Row]) => {
-      import CdpConnector._
-      val batches = rows.grouped(Constants.DefaultBatchSize).toVector
-
-      config.onConflict match {
-        case OnConflict.ABORT =>
-          batches.grouped(Constants.MaxConcurrentRequests).foreach { batchGroup =>
-            batchGroup.parTraverse(relation.insert).unsafeRunSync()
-          }
-
-        case OnConflict.UPSERT =>
-          batches.grouped(Constants.MaxConcurrentRequests).foreach { batchGroup =>
-            batchGroup.parTraverse(relation.upsert).unsafeRunSync()
-          }
-
-        case OnConflict.UPDATE =>
-          batches.grouped(Constants.MaxConcurrentRequests).foreach { batchGroup =>
-            batchGroup.parTraverse(relation.update).unsafeRunSync()
-          }
-
-        case OnConflict.DELETE =>
-          batches.grouped(Constants.MaxConcurrentRequests).foreach { batchGroup =>
-            batchGroup.parTraverse(relation.delete).unsafeRunSync()
-          }
+    if (resourceType == "assetshierarchy") {
```

Comment:
Still don't see why we couldn't just make this part of the match? Apologies if I missed your explanation on the previous review.

Reply:
There's probably a way, but the Asset hierarchy builder needs to work on the entire dataset, not a batched subset. We could move the batching into each of the implementations of the `WritetableResource` methods for the other resources, but I don't think it's worth it.

Reply:
Ah, I missed the `data.foreachPartition` part below the match. Alrightie then.